File: 2019/01/haskell/day_01.hs
---------------------------------------------------------------------
-- https://adventofcode.com/2019/day/1
---------------------------------------------------------------------

-- $
-- >>> fuel <$> [12,14,1969,100756]
-- [2,2,654,33583]
fuel :: Int -> Int
fuel i = (i `div` 3) - 2

part_1 :: [Int] -> Int
part_1 = sum . map fuel

part_2 :: [Int] -> Int
part_2 = sum . map (sum . takeWhile (> 0) . tail . iterate fuel)

main :: IO ()
main = do
  xs <- map read . words <$> getContents :: IO [Int]
  print $ part_1 xs
  print $ part_2 xs

File: 2019/python/day_01.py
"""
https://adventofcode.com/2019/day/1
"""

from typing import List


def get_lines():
    xs = []

    def loop():
        try:
            i = int(input())
            xs.append(i)
            loop()
        except EOFError:
            return

    loop()
    return xs


def fuel(i: int) -> int:
    """
    >>> [fuel(i) for i in [12,14,1969,100756]]
    [2, 2, 654, 33583]
    """
    return i // 3 - 2


def part_1(xs: List[int]) -> int:
    return sum([fuel(x) for x in xs])


def part_2(xs: List[int]) -> int:
    ss = []
    for x in xs:
        s = fuel(x)
        while s > 0:
            ss.append(s)
            s = fuel(s)

    return sum(ss)


def main():
    xs = get_lines()
    print(part_1(xs))
    print(part_2(xs))


if __name__ == '__main__':
    main()
